docs/User-guide.rst

```diff
@@ -0,0 +1,73 @@
+**********
+User Guide
+**********
+
+Once you understand :ref:`how dedupe works <how-it-works-label>`, and you've taken
+a look at some of the :doc:`Examples`,
+you probably want to figure out how to adapt dedupe to your own problem. This
+user guide is for that.
+
+Memory Considerations
+=====================
+
+The most common memory bottleneck is during clustering. Within every block,
+after computing a score for each pair of records, dedupe then runs a
+connected-components algorithm as a pre-processing step for the hierarchical
+agglomerative clustering.
+`Currently, this requires having the entire edge list in memory
+<https://github.com/dedupeio/dedupe/issues/819>`__, which for :math:`N`` records
+leads to :math:`O(N^2)` memory usage. If you end up with any "superblocks" because
+your blocking rules weren't picky enough, this can run you out of memory. You
+could inspect your blocks prior to clustering, and possibly split them up further.
+
+The other major memory bottleneck is building the index for blocking. If this is
+a problem, you can try turning off index blocking rules (and just use predicate
+blocking rules) by setting ``index_predicates=False`` in
+:meth:`dedupe.Dedupe.train`.
+
+Time Considerations
+===================
+
+Similar to the memory considerations, the slow part of dedupe is whenever
```

Comment:
i think clustering was a major bottleneck until https://github.com/dedupeio/dedupe/pull/912, i doubt it's where the problem is now.

i would expect that blocking can be pretty slow depending on how many predicates you learned and what type.

Reply:
OK I can swap it to this. In that case, the advice for dealing with slow blocking would be:
- Turn off index predicates
- inspect what predicates you learned. If a lot (>20??) of them, then you should...
  - Remove some of your training samples?
  - Reduce `recall` in your call to `train()`?

Is that correct?

Reply:
yeah, i would recommend reducing the number of positive examples in the the training data and turn off index predicates. i would make sure that people understand they are going to be trading off recall for speed probably.

this is also an area where can probably benefit from parallelization or other code strategies, although current attempts haven't really proved promising yet.